### PR TITLE
Add constructing TinyAsciiStr from utf16

### PR DESCRIPTION
**Pull request by zbraniecki**
Introduces TinyAsciiStr constructors from utf16 and converges on the consensus from #4931.

### CODE REVIEW

**Comment by zbraniecki** [comment]
@sffc can I get feedback before I port all of icu4x to new names?

File: utils/tinystr/src/ascii.rs [hunk 1]
```diff
@@ -16,20 +16,25 @@ pub struct TinyAsciiStr<const N: usize> {
 }
 
 impl<const N: usize> TinyAsciiStr<N> {
-    /// Creates a `TinyAsciiStr<N>` from the given byte slice.
+    /// Creates a `TinyAsciiStr<N>` from the given utf8 byte slice.
     /// `bytes` may contain at most `N` non-null ASCII bytes.
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:20]:
```suggestion
    /// Creates a `TinyAsciiStr<N>` from the given UTF-8 slice.
    /// `bytes` may contain at most `N` non-null ASCII code points.
```

File: utils/tinystr/src/ascii.rs [hunk 2]
```diff
@@ -16,20 +16,25 @@ pub struct TinyAsciiStr<const N: usize> {
...
-    pub const fn from_bytes(bytes: &[u8]) -> Result<Self, TinyStrError> {
-        Self::from_bytes_inner(bytes, 0, bytes.len(), false)
+    pub const fn try_from_utf8(bytes: &[u8]) -> Result<Self, TinyStrError> {
+        Self::try_from_utf8_inner(bytes, 0, bytes.len(), false)
     }
 
-    /// Creates a `TinyAsciiStr<N>` from a byte slice, replacing invalid bytes.
+    /// Creates a `TinyAsciiStr<N>` from the given utf16 byte slice.
+    /// `bytes` may contain at most `N` non-null ASCII bytes.
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:26]:
```suggestion
    /// Creates a `TinyAsciiStr<N>` from the given UTF-16 slice.
    /// `bytes` may contain at most `N` non-null ASCII code points.
```

File: utils/tinystr/src/ascii.rs [hunk 3]
```diff
@@ -16,20 +16,25 @@ pub struct TinyAsciiStr<const N: usize> {
...
+    pub const fn try_from_utf16(bytes: &[u16]) -> Result<Self, TinyStrError> {
+        Self::try_from_utf16_inner(bytes, 0, bytes.len(), false)
+    }
+
+    /// Creates a `TinyAsciiStr<N>` from a utf8 byte slice, replacing invalid bytes.
     ///
     /// Null and non-ASCII bytes (i.e. those outside the range `0x01..=0x7F`)
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:33]:
```suggestion
    /// Creates a `TinyAsciiStr<N>` from a UTF-8 slice, replacing invalid code units.
    ///
    /// Invalid code units, as well as null or non-ASCII code points (i.e. those outside the range U+0001..=U+007F`)
```

File: utils/tinystr/src/ascii.rs [hunk 4]
```diff
@@ -42,7 +47,37 @@ impl<const N: usize> TinyAsciiStr<N> {
...
+
+        Self {
+            // SAFETY: `out` only contains ASCII bytes and has same size as `self.bytes`
+            bytes: unsafe { AsciiByte::to_ascii_byte_array(&out) },
+        }
+    }
+
+    /// Creates a `TinyAsciiStr<N>` from a utf16 byte slice, replacing invalid bytes.
+    ///
+    /// Null and non-ASCII bytes (i.e. those outside the range `0x01..=0x7F`)
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:63]:
```suggestion
    /// Creates a `TinyAsciiStr<N>` from a UTF-16 slice, replacing invalid code units.
    ///
    /// Invalid code units, as well as null or non-ASCII code points (i.e. those outside the range U+0001..=U+007F`)
```

File: utils/tinystr/src/ascii.rs [hunk 5]
```diff
@@ -16,20 +16,25 @@ pub struct TinyAsciiStr<const N: usize> {
...
-    pub const fn from_bytes(bytes: &[u8]) -> Result<Self, TinyStrError> {
-        Self::from_bytes_inner(bytes, 0, bytes.len(), false)
+    pub const fn try_from_utf8(bytes: &[u8]) -> Result<Self, TinyStrError> {
+        Self::try_from_utf8_inner(bytes, 0, bytes.len(), false)
     }
 
-    /// Creates a `TinyAsciiStr<N>` from a byte slice, replacing invalid bytes.
+    /// Creates a `TinyAsciiStr<N>` from the given utf16 byte slice.
+    /// `bytes` may contain at most `N` non-null ASCII bytes.
+    pub const fn try_from_utf16(bytes: &[u16]) -> Result<Self, TinyStrError> {
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:27]: `bytes` is not a correct name for a `[u16]`. `utf16` or `code_units` as well

File: utils/tinystr/src/ascii.rs [hunk 6]
```diff
@@ -16,20 +16,25 @@ pub struct TinyAsciiStr<const N: usize> {
 }
 
 impl<const N: usize> TinyAsciiStr<N> {
-    /// Creates a `TinyAsciiStr<N>` from the given byte slice.
+    /// Creates a `TinyAsciiStr<N>` from the given utf8 byte slice.
     /// `bytes` may contain at most `N` non-null ASCII bytes.
-    pub const fn from_bytes(bytes: &[u8]) -> Result<Self, TinyStrError> {
-        Self::from_bytes_inner(bytes, 0, bytes.len(), false)
+    pub const fn try_from_utf8(bytes: &[u8]) -> Result<Self, TinyStrError> {
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:21]: Consider naming this argument `utf8` or `code_units`.

**Review by robertbastian — COMMENTED** [review]
Some naming observations

**Comment by zbraniecki** [comment]
Thanks! Applied!

File: utils/tinystr/src/ascii.rs [hunk 7]
```diff
@@ -22,16 +22,17 @@ impl<const N: usize> TinyAsciiStr<N> {
         Self::try_from_utf8_inner(bytes, 0, bytes.len(), false)
     }
 
-    /// Creates a `TinyAsciiStr<N>` from the given utf16 byte slice.
-    /// `bytes` may contain at most `N` non-null ASCII bytes.
-    pub const fn try_from_utf16(bytes: &[u16]) -> Result<Self, TinyStrError> {
-        Self::try_from_utf16_inner(bytes, 0, bytes.len(), false)
+    /// Creates a `TinyAsciiStr<N>` from the given UTF-16 slice.
+    /// `bytes` may contain at most `N` non-null ASCII code points.
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:26]:
```suggestion
    /// `code_points` may contain at most `N` non-null ASCII code points.
```

File: utils/tinystr/src/ascii.rs [hunk 8]
```diff
@@ -22,16 +22,17 @@ impl<const N: usize> TinyAsciiStr<N> {
...
     }
 
-    /// Creates a `TinyAsciiStr<N>` from a utf8 byte slice, replacing invalid bytes.
+    /// Creates a `TinyAsciiStr<N>` from a UTF-8 slice, replacing invalid code units.
     ///
-    /// Null and non-ASCII bytes (i.e. those outside the range `0x01..=0x7F`)
-    /// will be replaced with the '?' character.
+    /// Invalid code units, as well as null or non-ASCII code points
+    /// (i.e. those outside the range U+0001..=U+007F`)
+    /// will be replaced with the replacement byte.
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:35]: it think it makes sense to document that the replacement character is `?`

Comment (sffc) [reply utils/tinystr/src/ascii.rs:35]: The replacement character is an argument to the function

File: utils/tinystr/src/ascii.rs [hunk 9]
```diff
@@ -16,20 +16,26 @@ pub struct TinyAsciiStr<const N: usize> {
...
     ///
     /// The input slice will be truncated if its length exceeds `N`.
-    pub const fn from_bytes_lossy(bytes: &[u8]) -> Self {
-        const QUESTION: u8 = b'?';
+    pub const fn try_from_utf8_lossy(bytes: &[u8], replacement: u8) -> Self {
```
Comment (sffc) [inline utils/tinystr/src/ascii.rs:38]: These do not return Result so they shouldn't be called "try"

File: utils/tinystr/src/ascii.rs (hunk 8 above)
Comment (robertbastian) [reply utils/tinystr/src/ascii.rs:35]: oh didn't notice that change, nevermind then

**Review by Manishearth — DISMISSED** [review]
i like this

**Comment by zbraniecki** [comment]
Btw. the perf numbers I'm getting from this work for UTF-16 strings is ~10x improvement for construction of TinyStr, and for construction of subtags on top of them. The resulting perf is on par between UTF-8 and UTF-16 and I expect those numbers to carry up to LanguageIdentifier and Locale.

**Comment by sffc** [comment]
> Btw. the perf numbers I'm getting from this work for UTF-16 strings is ~10x improvement for construction of TinyStr, and for construction of subtags on top of them (compared to converting a Vec to a String). The resulting perf is on par between UTF-8 and UTF-16 and I expect those numbers to carry up to LanguageIdentifier and Locale.

Great; I was skeptical that this would be much of an improvement, but if it works that's great. What are the specific numbers you get from `construct/utf8/TinyAsciiStr` and `construct/utf16/TinyAsciiStr` in this PR?

**Comment by zbraniecki** [comment]
I like to think I wouldn't spend so much time on this without an early signal that the perf win is substantial, but... ;)

Here are the numbers (mbp m1):

```
overview/construct/utf8/TinyAsciiStr
                        time:   [272.35 ns 272.92 ns 273.53 ns]
overview/construct/utf16-from-utf8/TinyAsciiStr
                        time:   [2.5409 µs 2.5495 µs 2.5596 µs]
overview/construct/utf16/TinyAsciiStr
                        time:   [272.97 ns 273.90 ns 274.84 ns]
```

**Comment by sffc** [comment]
Where's the code for `overview/construct/utf16-from-utf8/TinyAsciiStr`?

Just trying to keep you honest :)

**Comment by sffc** [comment]
I guess what I'm thinking is that if you are using a `String` in your benches, then of course it will be slower, but you shouldn't need to use a `String`. You could walk the string once, put it in a stack buffer, and then feed the stack buffer to TinyStr.

That said, it's really annoying to do that in user land, so I am sold on this PR if simply for the ergonomics.

File: utils/tinystr/src/ascii.rs [hunk 10]
```diff
@@ -131,16 +182,59 @@ impl<const N: usize> TinyAsciiStr<N> {
         })
     }
 
-    // TODO: This function shadows the FromStr trait. Rename?
     #[inline]
-    pub const fn from_str(s: &str) -> Result<Self, TinyStrError> {
-        Self::from_bytes_inner(s.as_bytes(), 0, s.len(), false)
+    pub(crate) const fn try_from_utf16_inner(
+        code_points: &[u16],
```
Comment (sffc) [inline utils/tinystr/src/ascii.rs:187]: Thought/Suggestion (optional): You don't really need the UTF-16 version to be `const` so you could implement this to instead work from `Iterator<char>` which would be convenient for other string sources like UTF-32 (python) or reader streams.

**Review by sffc — APPROVED** [review]

**Comment by zbraniecki** [comment]
> Where's the code for overview/construct/utf16-from-utf8/TinyAsciiStr?

I didn't land it since I don't think it is worth keeping comparison to a non-chosen path.

> I guess what I'm thinking is that if you are using a String in your benches, then of course it will be slower, but you shouldn't need to use a String. 

Yes, the slow path is exactly what Boa is doing when [constructing a Locale](https://github.com/boa-dev/boa/blob/95ba66094364680b53da403bf63549a38d99c565/core/engine/src/builtins/intl/locale/mod.rs#L215C18-L215C39) from [JsValue](https://github.com/boa-dev/boa/blob/95ba66094364680b53da403bf63549a38d99c565/core/string/src/lib.rs#L1121-L1131).

> You could walk the string once, put it in a stack buffer, and then feed the stack buffer to TinyStr.

That would look nicely in the benchmark, but wouldn't represent real world JS engine scenario, would it?

File: utils/tinystr/src/ascii.rs (hunk 10 above)
Comment (zbraniecki) [reply utils/tinystr/src/ascii.rs:187]: I'm open to that in the future. I did a quick test and switching the `code_points` to take `Iterator<Item = u8>` cost ~5%. It's acceptable but I'd prefer to leave it for when we have a use case.

**Review by sffc — DISMISSED** [review]

**Pull request merged**